datafusion/physical-expr/src/expressions/with_filter.rs

```diff
@@ -0,0 +1,144 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::PhysicalExpr;
+use arrow::array::BooleanArray;
+use arrow::compute::filter_record_batch;
+use arrow::error::Result as ArrowResult;
+use arrow::record_batch::RecordBatch;
+use datafusion_common::DataFusionError;
+use std::fmt;
+use std::sync::Arc;
+
+#[derive(Debug)]
+pub struct ExprWithFilter {
```

Comment:
this is a neat idea

Can you please add some doc comments describing what this expression does?

I wonder if it will be a problem if this physical expr works differently from other physical exprs in that the output cardinality (number of rows) is different than the input cardinality?

Reply:
I've updated the doc string.

> I wonder if it will be a problem if this physical expr works differently from other physical exprs in that the output cardinality (number of rows) is different than the input cardinality?

 I think, it won't be a problem since filter is only applied for aggregation.

Reply:
Yess, you are right. 

Filtering rows causing issue while group by slicing. I've fixed it now